R/auth.R

```diff
@@ -0,0 +1,292 @@
+#' @param name Cache name to use.
+#' @export
+#' @examples 
+#' \dontrun{
+#' # save app auth for use in other sessions
+#' auth <- rtweet_app()
+#' auth_save(auth, "my-app")
+#' 
+#' # later, in a different session...
+#' auth_us("my-app")
+#' }
+auth_save <- function(auth, name) {
+  stopifnot(is_auth(auth), is_string(name))
+  
+  path <- auth_path(paste0(name, ".rds"))
+  inform(paste0("Saving auth to '", path, "'"))
+  
+  dir.create(auth_path(), showWarnings = FALSE, recursive = TRUE)
+  saveRDS(auth, path)
+  invisible(path)
+}
+
+auth_path <- function(...) {
+  # Use private option to make testing easier
+  path <- getOption("rtweet:::config_dir", rappdirs::user_config_dir("rtweet", "R"))
+  file.path(path, ...)
+}
+
+# Set default auth -------------------------------------------------------------
+
+#' Set default authentication
+#' 
+#' `auth_as()` sets up the default authentication mechanism used by all 
+#' rtweet API calls. See [rtweet_user()] to learn more about the three
+#' available authentication options.
+#' 
+#' @param auth One of the following options:
+#'   * `NULL`, the default, will look for cached authentication from a previous
+#'      session. If none are found, will automatically set up [rtweet_user()];
+#'      if one is found, will use that; if multiple are found, it will error.
+#'   * A string giving the name of a cached auth file.
+#'   * An object created by [rtweet_app()], [rtweet_bot()], or [rtweet_user()].
+#' @return Invisibly returns the previous authentication mechanism.
+#' @examples 
+#' \dontrun{
+#' # Use app auth for the remainder of this session:
+#' my_app <- rtweet_app()
+#' auth_as(my_app)
+#' 
+#' # Switch back to the default user based auth
+#' auth_as()
+#' 
+#' # Load auth saved by auth_save()
+#' auth_as("my-saved-app")
+#' }
+auth_as <- function(auth = NULL) {
+  old <- .state$auth
+  .state$auth <- find_auth(auth)
+  invisible(old)
+}
+
+find_auth <- function(auth = NULL) {
+  if (is.null(auth)) {
+    if (is_testing()) {
+      rtweet_test() %||% no_token()
+    } else if (is_dev_mode()) {
+      rtweet_test() %||% default_cached_auth()
+    } else{
+      default_cached_auth()
+    }
+  } else if (is_auth(auth)) {
+    auth
+  } else if (is_string(auth)) {
+    path <- auth_path(paste0(auth, ".rds"))
+    if (!file.exists(path)) {
+      abort(paste0("Can't find saved auth with name '", auth, "'"))
+    }
+    inform(paste0("Reading auth from '", path, "'"))
+    readRDS(path)
+  } else {
+    abort("Unrecognised input to `auth`")
+  }
+}
+
+default_cached_auth <- function() {
+  options <- dir(auth_path(), full.names = TRUE, pattern = "\\.rds$")
+  
+  if (length(options) == 0) {
+    if (is_interactive()) {
+      inform("No saved authentication found; creating...")
+      auth_save(rtweet_user(), "default")
+    } else {
+      abort("No saved auth found; please resolve in an interactive session")
+    }
+  } else if (length(options) == 1) {
+    readRDS(options)
+  } else {
```

Comment:
I'm fine with making that explicit; I only did that because I thought that's what you wanted.

Reply:
How about we have an explicit `auth_setup()` that does `auth_save(auth_user(), "default")`. Then this function would use `default.rds` if available, otherwise it would error with informative instructions.  

This means that you'd need to call `auth_setup()` once the first time you use rtweet 1.0.0.

Reply:
mmh, yes I want the default to be saved but knowing what I'm saving with an informative name, so that when using auth_as next time I know which token I'm using. Default is not enough informative because there are two auth defaults provided by rtweet: rtweet_app and rtweet_user.

Reply:
I'm not sure what you mean here — there is only one default auth provided by rtweet: `rtweet_user()`.